### PR TITLE
Add more documentation around layers.

### PR DESCRIPTION
**Pull request by Hixie**
cc @mjohnsullivan

### CODE REVIEW

File: packages/flutter/lib/src/rendering/layer.dart [hunk 1]
```diff
@@ -1691,7 +1704,10 @@ class ShaderMaskLayer extends ContainerLayer {
     }
   }
 
-  /// The size of the shader.
+  /// The position and size of the shader.
+  ///
+  /// The [shader] is only recendered inside this rectangle, using the top left
```
Comment (mjohnsullivan) [inline packages/flutter/lib/src/rendering/layer.dart]: 'recendered' -> 'rerendered'?

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -394,28 +403,39 @@ class PaintingContext extends ClipContext {
...
+  /// When `needsCompositing` is false, this method will use a more efficient
+  /// way to apply the layer effect than actually creating a layer.
+  /// {@endtemplate}
+  ///
+  /// {@template flutter.rendering.object.pushClipLayer.offset}
+  /// The `offset` argument is the offset from the origin of the canvas'
+  /// coordinate system to the origin of the caller's coordinate system.
+  /// {@endtemplate}
+  ///
+  /// The `clipRect` is rectangle (in the caller's coordinate system) to use to
```
Comment (mjohnsullivan) [inline packages/flutter/lib/src/rendering/object.dart]: 'is rectangle' -> 'is the rectangle'

**Comment by mjohnsullivan** [comment]
LGTM with comments

**Comment by Hixie** [comment]
Thanks for your review!

**Pull request merged**